### DISCUSSION TITLE
Pressure control implementation  in a hydrogen pipe network

### DISCUSSION THREAD

**Discussion by jdbelenol** [e2nIEE/pandapipes#666] [Q&A]
Dear Everyone,
I am currently working on a hydrogen network as depicted in Figure 1.

Figure 1

One of the requirements is that the pressure at the junction (J2), where the sink is connected, should be 30 bar. However, due to the pressure drop across the 100 km pipe (pipe 1), the current pressure is 15.93 bar instead.
To address this issue, I implemented a pressure control between junctions J2 and Jc2. The corresponding code for this pressure control is as follows:
import pandas as pd
import pandapipes as pp
import pandapipes.plotting as plot
from pandapipes import create_pressure_control

# Creating empty net
net_h2 = pp.create_empty_network(fluid="hydrogen")
# Creating junctions
j0 = pp.create_junction(net_h2, pn_bar=3.5, tfluid_k=293.15, in_service=True, type="junction", name="junction 0")
j1 = pp.create_junction(net_h2, pn_bar=30, tfluid_k=293.15, in_service=True, type="junction", name="junction 1")
j2 = pp.create_junction(net_h2, pn_bar=30, tfluid_k=293.15, in_service=True, type="junction", name="junction 2")
jc2 = pp.create_junction(net_h2, pn_bar=30, tfluid_k=293.15, in_service=True, type="junction", name="junction press control")

# Creating pressure control
press_control_1 = create_pressure_control(net_h2, from_junction=j2, to_junction=jc2, controlled_junction=jc2, controlled_p_bar=30, control_active=True, name="press_control_1")

# Creating pipes
p1 = pp.create_pipe_from_parameters(net_h2, from_junction=j1, to_junction=j2, length_km=100, diameter_m=0.3128, k_mm= 0.2, text_k = 293.15, in_service=True, type="pipe", name="pipe_1")  #to find out with kind of pipes are usually use for hydrogen production
# Creating external grid
h2_ext_grid = pp.create_ext_grid(net_h2, junction=j0, p_bar=3.5, t_k=293.15, in_service=True, type="pt", name="ext_grid")
# Creating sources
h2_source = pp.create_source(net_h2, junction=j0, mdot_kg_per_s=2.3871, scaling=1, in_service=True, type="source", name="source_1")
# Creating sinks
sink_1 = pp.create_sink(net_h2, junction=jc2, mdot_kg_per_s=2.3871, scaling=1, in_service=True,  type="sink", name="sink_1")
# Creating compressors
comp_1 = pp.create_compressor(net_h2, from_junction=j0, to_junction=j1, pressure_ratio=6.8889,  in_service=True, name="compressor_1")
# Running the pipeflow
pp.pipeflow(net_h2)
I would greatly appreciate it if you could review the implementation of the pressure control and confirm whether it has been done correctly. I am uncertain because, with the pressure control in place, I expected an increase in the pressure at J1 (the junction at the outlet of the compressor). Consequently, this should have resulted in an increase in the compressor power, given the higher outlet pressure required to overcome the increased pressure at J1 caused by the required pressure of 30 bar at J2. However, this did not happen - the pressure at J1 remains the same as before of the implementation of the pressure control, and, as result, the compressor power also remains the same. 
Additionally, I would be grateful for any insights on achieving the intended goal: ensuring that the compressor power is adjusted to maintain a pressure of 30 bar at the junction where the sink is connected, after implementing the pressure control (J2c).
Thank you in advance for your guidance and support.
Best regards,
Jose 
[Figure 1.pdf](https://github.com/user-attachments/files/17902300/Figure.1.pdf)

**Comment by dlohmeier** [comment]
Dear Jose,
sorry for the late reply, I have been absent for some time and probably my colleagues could not clarify this problem. The pressure controller is working slightly differently than you expected. It just creates an output pressure as desired, and injects a pressure lift or drop, depending on what you need, without calculating the required power or even temperature changes.
What you might want is the following: You could use a pressure controller instead of the compressor between J0 and J1 and set the controlled_junction to J2. This should lead to an injected pressure lift as expected. In order to calculate the required power for this pressure lift, you can use the formula given in pump_component.py:

    p_from = branch_results["p_abs_from"][f:t]
	p_to = branch_results["p_abs_to"][f:t]
	t0 = net["_pit"]["node"][from_nodes, TINIT_NODE]
	mf_sum_int = branch_results["mf_from"][f:t]
	# calculate ideal compression power
	compr = get_fluid(net).get_property("compressibility", p_from)
	try:
	    molar_mass = net.fluid.get_molar_mass()  # [g/mol]
	except UserWarning:
	    logger.error('Molar mass is missing in your fluid. Before you are able to '
		         'retrieve the compression power make sure that the molar mass is'
		         ' defined')
	else:
	    r_spec = 1e3 * R_UNIVERSAL / molar_mass  # [J/(kg * K)]
	    cp = net.fluid.get_heat_capacity(t0)
	    cv = cp - r_spec
	    k = cp/cv  # 'kappa' heat capacity ratio
	    w_real_isentr = (k / (k - 1)) * r_spec * compr * t0 * \
		            (np.divide(p_to, p_from) ** ((k - 1) / k) - 1)
	    res_table['compr_power_mw'].values[:] = \
		w_real_isentr * np.abs(mf_sum_int) / 1e6

You need to replace some values here, especially the pressure of inlet / outlet junction. Hope that helps!
